graphql_compiler/tests/test_compiler.py

```diff
@@ -5343,7 +5434,20 @@ def test_optional_and_traverse(self):
                 name: m.Animal___1.name
             ])}
         '''
-        expected_sql = NotImplementedError
+        expected_sql = '''
+            SELECT
+                [Animal_1].name AS child_name,
+                [Animal_2].name AS grandchild_name,
+                [Animal_3].name AS name
+            FROM
+                db_1.schema_1.[Animal] AS [Animal_3]
+                LEFT OUTER JOIN db_1.schema_1.[Animal] AS [Animal_1]
+                    ON [Animal_3].uuid = [Animal_1].parent
+                LEFT OUTER JOIN db_1.schema_1.[Animal] AS [Animal_2]
+                    ON [Animal_1].uuid = [Animal_2].parent
+            WHERE
+                [Animal_2].parent IS NOT NULL OR [Animal_1].parent IS NULL
```

Comment:
Let's imagine a situation where someone is running a SQL database without constraint checking. It might be possible that `[Animal_2].parent IS NOT NULL` is satisfied but the join to `Animal_3` still doesn't find anything. For the not-null case, might we be better off checking that the primary key on `Animal_3` is non-null?

Put another way, the foreign key column being `null` guarantees there's no edge there. But it being non-null only guarantees there's an edge there if constraint-checking is on, and deletes run with `CASCADE` set. While it's a good idea to do this, many databases and systems do not.

If you think it's too much work, it's fine either way. But if it's about equal work either way, let's err on the side of supporting more use cases.

Reply:
I don't get what you mean. We're not joining to `Animal_3`, that's where we start. Talk offline?

Reply:
Resolved offline. Just a misunderstanding.